File: go/datastructures/hash_table.go
package datastructures

import "hash/fnv"

// HashTable implements a hash table with open addressing and linear probing.
// Time: O(1) average for all operations
// Space: O(n)
type HashTable[K comparable, V any] struct {
	buckets  []*hashEntry[K, V]
	size     int
	capacity int
}

type hashEntry[K comparable, V any] struct {
	key     K
	value   V
	deleted bool
}

const loadFactorThreshold = 0.7

func NewHashTable[K comparable, V any](initialCapacity ...int) *HashTable[K, V] {
	cap := 16
	if len(initialCapacity) > 0 && initialCapacity[0] > 0 {
		cap = initialCapacity[0]
	}
	return &HashTable[K, V]{
		buckets:  make([]*hashEntry[K, V], cap),
		capacity: cap,
	}
}

func (h *HashTable[K, V]) Put(key K, value V) {
	if float64(h.size) >= float64(h.capacity)*loadFactorThreshold {
		h.resize(h.capacity * 2)
	}

	index := h.hash(key)
	firstDeleted := -1

	for i := 0; i < h.capacity; i++ {
		entry := h.buckets[index]

		if entry == nil {
			insertIdx := index
			if firstDeleted != -1 {
				insertIdx = firstDeleted
			}
			h.buckets[insertIdx] = &hashEntry[K, V]{key: key, value: value}
			h.size++
			return
		}

		if entry.deleted && firstDeleted == -1 {
			firstDeleted = index
		} else if !entry.deleted && entry.key == key {
			entry.value = value
			return
		}

		index = (index + 1) % h.capacity
	}

	if firstDeleted != -1 {
		h.buckets[firstDeleted] = &hashEntry[K, V]{key: key, value: value}
		h.size++
	}
}

func (h *HashTable[K, V]) Get(key K) (V, bool) {
	var zero V
	index := h.findIndex(key)
	if index == -1 {
		return zero, false
	}
	return h.buckets[index].value, true
}

func (h *HashTable[K, V]) Remove(key K) (V, bool) {
	var zero V
	index := h.findIndex(key)
	if index == -1 {
		return zero, false
	}
	entry := h.buckets[index]
	value := entry.value
	entry.deleted = true
	h.size--
	return value, true
}

func (h *HashTable[K, V]) Contains(key K) bool {
	return h.findIndex(key) != -1
}

func (h *HashTable[K, V]) IsEmpty() bool {
	return h.size == 0
}

func (h *HashTable[K, V]) Size() int {
	return h.size
}

func (h *HashTable[K, V]) Keys() []K {
	result := make([]K, 0, h.size)
	for _, entry := range h.buckets {
		if entry != nil && !entry.deleted {
			result = append(result, entry.key)
		}
	}
	return result
}

func (h *HashTable[K, V]) Values() []V {
	result := make([]V, 0, h.size)
	for _, entry := range h.buckets {
		if entry != nil && !entry.deleted {
			result = append(result, entry.value)
		}
	}
	return result
}

func (h *HashTable[K, V]) Clear() {
	h.buckets = make([]*hashEntry[K, V], h.capacity)
	h.size = 0
}

func (h *HashTable[K, V]) hash(key K) int {
	hasher := fnv.New64a()
	hasher.Write([]byte(any(key).(string)))
	return int(hasher.Sum64() % uint64(h.capacity))
}

func (h *HashTable[K, V]) findIndex(key K) int {
	index := h.hash(key)

	for i := 0; i < h.capacity; i++ {
		entry := h.buckets[index]

		if entry == nil {
			return -1
		}

		if !entry.deleted && entry.key == key {
			return index
		}

		index = (index + 1) % h.capacity
	}

	return -1
}

func (h *HashTable[K, V]) resize(newCapacity int) {
	oldBuckets := h.buckets
	h.buckets = make([]*hashEntry[K, V], newCapacity)
	h.capacity = newCapacity
	h.size = 0

	for _, entry := range oldBuckets {
		if entry != nil && !entry.deleted {
			h.Put(entry.key, entry.value)
		}
	}
}

File: java/src/main/java/com/dsa/datastructures/HashTable.java
package com.dsa.datastructures;

import java.util.*;

/**
 * A hash table implementation using open addressing with linear probing.
 * Time: O(1) average for insert, get, remove
 * Space: O(n)
 */
public class HashTable<K, V> {
    private static final int INITIAL_CAPACITY = 16;
    private static final double LOAD_FACTOR = 0.75;

    private enum EntryState { EMPTY, DELETED, OCCUPIED }

    private static class Entry<K, V> {
        K key;
        V value;
        EntryState state;

        Entry() {
            this.state = EntryState.EMPTY;
        }

        void set(K key, V value) {
            this.key = key;
            this.value = value;
            this.state = EntryState.OCCUPIED;
        }

        void delete() {
            this.key = null;
            this.value = null;
            this.state = EntryState.DELETED;
        }
    }

    private Entry<K, V>[] entries;
    private int size;
    private int capacity;

    @SuppressWarnings("unchecked")
    public HashTable() {
        this.capacity = INITIAL_CAPACITY;
        this.entries = new Entry[capacity];
        for (int i = 0; i < capacity; i++) {
            entries[i] = new Entry<>();
        }
        this.size = 0;
    }

    private int hash(K key) {
        return Math.abs(key.hashCode()) % capacity;
    }

    @SuppressWarnings("unchecked")
    private void resize() {
        Entry<K, V>[] oldEntries = entries;
        capacity *= 2;
        entries = new Entry[capacity];
        for (int i = 0; i < capacity; i++) {
            entries[i] = new Entry<>();
        }
        size = 0;

        for (Entry<K, V> entry : oldEntries) {
            if (entry.state == EntryState.OCCUPIED) {
                put(entry.key, entry.value);
            }
        }
    }

    public Optional<V> put(K key, V value) {
        if ((size + 1.0) / capacity > LOAD_FACTOR) {
            resize();
        }

        int index = hash(key);
        int startIndex = index;

        while (true) {
            Entry<K, V> entry = entries[index];

            if (entry.state == EntryState.EMPTY || entry.state == EntryState.DELETED) {
                entry.set(key, value);
                size++;
                return Optional.empty();
            }

            if (entry.state == EntryState.OCCUPIED && entry.key.equals(key)) {
                V oldValue = entry.value;
                entry.value = value;
                return Optional.of(oldValue);
            }

            index = (index + 1) % capacity;
            if (index == startIndex) {
                throw new RuntimeException("Hash table is full");
            }
        }
    }

    public Optional<V> get(K key) {
        int index = hash(key);
        int startIndex = index;

        while (true) {
            Entry<K, V> entry = entries[index];

            if (entry.state == EntryState.EMPTY) {
                return Optional.empty();
            }

            if (entry.state == EntryState.OCCUPIED && entry.key.equals(key)) {
                return Optional.of(entry.value);
            }

            index = (index + 1) % capacity;
            if (index == startIndex) {
                return Optional.empty();
            }
        }
    }

    public Optional<V> remove(K key) {
        int index = hash(key);
        int startIndex = index;

        while (true) {
            Entry<K, V> entry = entries[index];

            if (entry.state == EntryState.EMPTY) {
                return Optional.empty();
            }

            if (entry.state == EntryState.OCCUPIED && entry.key.equals(key)) {
                V value = entry.value;
                entry.delete();
                size--;
                return Optional.of(value);
            }

            index = (index + 1) % capacity;
            if (index == startIndex) {
                return Optional.empty();
            }
        }
    }

    public boolean containsKey(K key) {
        return get(key).isPresent();
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void clear() {
        for (int i = 0; i < capacity; i++) {
            entries[i] = new Entry<>();
        }
        size = 0;
    }
}
